Fix linkToVirtual's dispatch of invokeBasic

fengxue-IS opened the pull request:
Use the MemberName MethodType to compute the expected slotCount for
invoke target. Use J9_BCLOOP_SEND_TARGET_METHODHANDLE_INVOKEBASIC
to recognize invokeBasic call and inline the invokeBasic operation within linkToVirtual
to avoid setting temporary value and flags during VM transition.

Related: #13094 

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

fengxue-IS commented:
This PR fixes `java/lang/invoke/LoopCombinatorLongSignatureTest.java`
Testing with personal build, will remove WIP once testing passed

fengxue-IS commented:
FYI @tajila

tajila commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -8406,7 +8408,23 @@ class INTERPRETER_CLASS
+			/* +1 to account for the receiver */
+			methodArgCount = VM_VMHelpers::methodTypeParameterSlotCount(_currentThread, methodType) + 1;
+
+			if (J9_CLASS_FROM_METHOD(methodID->method) == J9VMJAVALANGINVOKEMETHODHANDLE(_currentThread->javaVM)) {
```
you can use `_vm` here

tajila replied:
should use the `J9VMJAVALANGINVOKEMETHODHANDLE_OR_NULL` instead of `J9VMJAVALANGINVOKEMETHODHANDLE`. AFAIK at this point we should be guaranteed that MH is loaded

tajila commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -8299,9 +8299,11 @@ class INTERPRETER_CLASS
 	{
 		VM_BytecodeAction rc = GOTO_RUN_METHOD;
 		bool fromJIT = J9_ARE_ANY_BITS_SET(_currentThread->jitStackFrameFlags, J9_SSF_JIT_NATIVE_TRANSITION_FRAME);
+		bool fromNF = J9_ARE_ANY_BITS_SET(_currentThread->privateFlags2, J9_PRIVATE_FLAGS2_NF_INVOKE_BASIC);
```
what does NF stand for?

fengxue-IS replied:
NamedFunction, it is the wrapper used to hold a `invokeBasic` MemberName

tajila commented on runtime/oti/j9consts.h:
```diff
@@ -178,6 +178,7 @@ extern "C" {
 #define J9_PRIVATE_FLAGS2_CHECK_PACKAGE_ACCESS 0x2
 #define J9_PRIVATE_FLAGS2_UNSAFE_HANDLE_SIGBUS 0x4
 #define J9_PRIVATE_FLAGS2_ASYNC_GET_CALL_TRACE 0x8 /* DO NOT set this anywhere but AsyncGetCallTrace */
+#define J9_PRIVATE_FLAGS2_NF_INVOKE_BASIC 0x10
```
Flag may not be necessary. Since you just need to track temporary state, you could just use another temp slot.

DanHeidinga reviewed (changes requested):
See comment for concerns with this approach.

Can we encode this differently?  Maybe aways set the code up to look for the methodArgCount in the `_currentThread->tempSlot` so there's no special case?

Or introduce a second form of `VMINLINE VM_BytecodeAction invokeBasic(REGISTER_ARGS_LIST)` - basically another vm entrypoint - to handle the "look in tempSlot for argCount" version?

There has to be a better factoring of this to avoid the use of privateFlags which will always get out of sync.

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -8433,6 +8442,17 @@ class INTERPRETER_CLASS
+		) {
+			j9object_t lambdaForm = J9VMJAVALANGINVOKEMETHODHANDLE_FORM(_currentThread, receiverObject);
+			j9object_t memberName = J9VMJAVALANGINVOKELAMBDAFORM_VMENTRY(_currentThread, lambdaForm);
+			_sendMethod = (J9Method *)(UDATA)J9OBJECT_U64_LOAD(_currentThread, memberName, _vm->vmtargetOffset);
```
I'm having trouble saying this is ok or not because I'm not clear on why the method being varargs matters when calling `invokeBasic`.  Is it related to `invokeBasic` being signaturePolymorphic?

Since this is for varargs, is there a need to collect the varargs into the trailing array here?

Sorry, I'm not clear why this code needs to handle varargs specially when the implementation of the `invokeBasic` VM_BytecodeAction doesn't.

fengxue-IS replied:
>Is it related to invokeBasic being signaturePolymorphic?

Yes, I guess I should have used a more exact check instead of `J9AccVarArgs` to avoid including regular varargs method which doesn't require the extra `methodTypeParameterSlotCount` call.

The VM_BytecodeAction `invokeBasic` impl reads the methodArgCount from ramCP which is based on the call signature in CP hence it knows where to locate the receiver object on the stack to load the target `MemberName` from

With methods called by `linkToVirtual`, since transition to new target is still on the same bytecode
(ie. `invokeStatic MethodHandle.linkToVirtual(...)`) the ramCP's methodArgCount would hold the number of parameters passed to `linkToVirtual` which doesn't match what target method actually received. So when looking for the receiver object, we used the `romMethod->ArgCount`

But because `invokeBasic` have a `romMethod->argCount` that may not be consistent with the call signature. the code have to find the actual number of argument expected through the signature from target MemberName's methodType.

DanHeidinga replied:
So this is the case where we have a `linkToVirtual` redispatching to `invokeBasic`?  

And I assume this is created by something like `MethodHandles.lookup().findVirtual(MethodHandle. invokeBasic, MethodType.methodType(...))`?  If that's the case, shouldn't we recognize that pattern on creation and return the `invokeBasic` directly rather than the `linkToVirtual` handle?

In our original implementation, we had special MH subclass to cover the signaturePolymorphic methods called `InvokeExactHandle` & `InvokeGenericHandle`.  I'd expect the OpenJDK impl to shortcut these cases in the resolve of the memberName as well.

fengxue-IS replied:
this is generated when the MH invoke arity is 254, since max JVM argCount is 255, using a invokeHandle instruction would not be possible as it adds extra MemberName and Appendix object to the argument list.

DanHeidinga replied:
So we're dealing with an extreme corner case here?  All that's critical info to put into the source code so we can understand the **why** behind the design.  If you include that info in a comment up front we don't have to try and reverse engineer it when reviewing the change.... and when you have to fix a bug in 6 months you'll be happy to have the info directly in the code.

Can you update this with a comment that describes that this is an uncommon case and why it has to be handled this way?

fengxue-IS replied:
I will add a comment to describe this special case and the reasons for such handling

DanHeidinga replied:
Can you also add a comment to the `invokeBasic` bytecodeAction to indicate that the code has been inlined here?  It'll make it easier to keep both copies in sync if they refer to each other

DanHeidinga commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -8406,7 +8406,16 @@ class INTERPRETER_CLASS
+		UDATA methodArgCount = 0;
+		bool isVararg = J9_ARE_ALL_BITS_SET(romMethod->modifiers, J9AccVarArgs);
+
+		if (isVararg) {
```
So this shouldn't be `isVarags` but something like `isSignaturePolymorphic`?  And varargs is an inexpensive proxy?

fengxue-IS commented:
added comments do describe design, personal build passed for sanity/extended and ` java/lang/invoke/LoopCombinatorLongSignatureTest.java`
Updated code to only check for `invokeBasic` as that is the only signature polymorphic method linked as a NamedFunction.

DanHeidinga commented:
Jenkins test sanity win,xlinux jdk16
